src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+
+            List<CodeAction> fixes = new List<CodeAction>();
+
+            var codeFixers = GetAllCodeFixers(analyzerAssemblies);
+
+            foreach (var item in diagnostics)
+            {
+                foreach (var codeFixer in codeFixers.GetValueOrDefault(item.Id, ImmutableList.Create<CodeFixProvider>()))
+                {
+                    fixes.AddRange(await GetFixesAsync(solution, codeFixer, item, cancellationToken).ConfigureAwait(false));
+                }
+            }
+
+            Console.WriteLine($"Found {fixes.Count} potential code fixes");
+
+            Console.WriteLine("Calculating changes");
+
+            stopwatch.Restart();
+
+            object lockObject = new object();
+
+            Parallel.ForEach(fixes, fix =>
+            {
+                try
+                {
+                    fix.GetOperationsAsync(cancellationToken).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    // Report thrown exceptions
+                    lock (lockObject)
+                    {
+                        Utilities.WriteLine($"The fix '{fix.Title}' threw an exception:", ConsoleColor.Yellow);
+                        Utilities.WriteLine(ex.ToString(), ConsoleColor.Red);
+                    }
+                }
+            });
+
+            Console.WriteLine($"Calculating changes completed in {stopwatch.ElapsedMilliseconds}ms");
+        }
+
+        private static async Task<IEnumerable<CodeAction>> GetFixesAsync(Solution solution, CodeFixProvider codeFixProvider, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            List<CodeAction> codeActions = new List<CodeAction>();
+
+            await codeFixProvider.RegisterCodeFixesAsync(new CodeFixContext(solution.GetDocument(diagnostic.Location.SourceTree), diagnostic, (a, d) => codeActions.Add(a), cancellationToken)).ConfigureAwait(false);
+
+            return codeActions;
+        }
+
+        private static Task<Statistic> GetAnalyzerStatisticsAsync(IEnumerable<Project> projects, CancellationToken cancellationToken)
+        {
+            ConcurrentBag<Statistic> sums = new ConcurrentBag<Statistic>();
+
+            Parallel.ForEach(projects.SelectMany(i => i.Documents), document =>
+            {
+                var documentStatistics = GetAnalyzerStatisticsAsync(document, cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult();
+                sums.Add(documentStatistics);
+            });
+
+            Statistic sum = sums.Aggregate(new Statistic(0, 0, 0), (currentResult, value) => currentResult + value);
+            return Task.FromResult(sum);
+        }
+
+        private static async Task<Statistic> GetAnalyzerStatisticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            SyntaxTree tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
if you don't want Async version, I think we actually have synchronous version of GetSyntaxRoot? I might be wrong though.

Reply:
I could not find one. TryGetSyntaxTree also does not support cancellation tokens.

---
In reply to: [150121042](https://github.com/dotnet/roslyn/pull/23087#discussion_r150121042) [](ancestors = 150121042)